library/core/src/mem/manually_drop.rs

```diff
@@ -40,9 +39,115 @@ use crate::ptr;
 /// }
 /// ```
 ///
+/// # Interaction with `Box`
+///
+/// Currently, once the `Box<T>` inside a `ManuallyDrop<Box<T>>` is dropped,
+/// moving the `ManuallyDrop<Box<T>>` is [considered to be undefined
```

Comment:
I think documentation should clarify that this also applies to any types that directly contain `Box<_>`, like `(Box<T>, u8)` or `struct Foo(Box<T>);`.

I also believe that per current rules, calling `ManuallyDrop::drop()` on any type that directly contains `Box<_>` is insta-UB, even if you don’t move it afterwards. Technically, it doesn’t neccessarily count as producing an invalid value:

> “Producing” a value happens any time a value is assigned to or read from a place, passed to a function/primitive operation or returned from a function/primitive operation.

but at least touching it in any way (like creating a reference) is certainly UB.

Reply:
If you don't name the place after calling `ManuallyDrop::drop`, everything is fine; the place isn't touched after that point. Because `ManuallyDrop` does not have any drop glue, it does not get a `drop_in_place` terminator at the end of scope, and thus the place is never "used" implicitly by the language, and no reference is created unless you create it yourself.

Reply:
I have reworded the text to clarify that the UB also applies to types containing `Box`.

I am pretty sure that calling `ManuallyDrop::drop()` on a type containing `Box` is *not* UB. https://github.com/rust-lang/unsafe-code-guidelines/issues/245 only states that using the `ManuallyDrop` later is UB. If you believe that `ManuallyDrop::drop()` is insta-UB, then I would like a citation to prove otherwise.

Of note, Miri doesn't seem to complain anything if I create a reference to an already-dropped `ManuallyDrop<Box<T>>`. I believe this is because we are still undecided on whether the existence of a `&T` implies that the `T` must be valid.

Reply:
Re-reading a reference, I don’t think I can prove that it’s insta-UB. This is certainly UB per reference though:

```rust
let mut x = ManuallyDrop::new(Box::new(42));
unsafe { ManuallyDrop::drop(&mut x); }
let _y = &x;
```

since it produces (by writing to a place) an invalid value. Reference takes a conservative stance and declares references to invalid values to be invalid:

> A reference or Box<T> that is [dangling](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers), misaligned, or points to an invalid value (in case of dynamically sized types, using the actual dynamic type of the pointee as determined by the metadata).

Reply:
From https://github.com/rust-lang/unsafe-code-guidelines/issues/412:
> Note that the Rust reference currently answers ["whether a reference requires the pointed-to data to be valid"] with "yes", but in my view this is mostly because we haven't yet figured out what exactly the weaker requirement is that we actually want to impose.

Reply:
Yeah, I know. I still think that documentation should be coherent: if we document this as being unsound in reference, it should be also documented as being unsound in other places that deal with the same situation.

Reply:
I just found out that the [nightly version of the reference](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html#invalid-values) now states:

> A reference or `Box<T>` must be aligned, it cannot be dangling, and it must point to a valid value (in case of dynamically sized types, using the actual dynamic type of the pointee as determined by the metadata). Note that the last point (about pointing to a valid value) remains a subject of some debate.